keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+-->
+
+Allow users to resize the volumes associated with `StatefulSet` by modifying the size of
+PVC in the `.spec.volumeClaimTemplates`.
+
+* Minimize the amount of unexpected errors/failures that the user might experience during volume resize by having as much validation as we can on the API server.
+* In case errors do happen during resize, we shouldn't get in the way of the user trying to remediate those errors
+* Provide a feedback mechanism for when a volume finished resizing successfully, so that users can build detection mechanisms around that for their automated CI/CD processes.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+This proposal does not try to address the following scenarios but relies on the
+underlying PVC resize logic to take the appropriate action or indicate error as needed.
+
+- Shrinking of volume is not allowed.
+- Few environments require Pod restarts in order to get access to the resized volume (meaning that the underlying CSI driver doesn't support online expansions). This KEP won't address such scenarios.
+- Modifying anything but the `resources.request.storage` field of the `persistentVolumeClaim` (eg, `storageClassName`, `labels`, `annotations`).
+- Resizing a PVC associated with a StatefulSet directly by modifying the PVC spec will not affect the StatefulSet `.spec.volumeClaimTemplates`.
+- Resizing ephemeral volumes.
+- Correcting failures resulting from bad input or environmental issues (such as insufficient capacity/quota). We'll make an attempt to do as much validation as we can in the API server, but cases of errors/failures past admission will have to be handled by the user. This KEP will discuss the different failures that could happen past admission.
+
+  
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Before we proceed to the proposal it is imperative to layout the various components involved
+in creating and managing the persistent volumes associated with `StatefulSet`. 
+
+- StatefulSet controller creates the PVCs from `volumeClaimTemplates` and associates them with the pods
+spawned from it. Reconciliation logic runs for every change made to the StatefulSet
+`Spec` and `Status` reflects the progress towards making the real world state equal to the
+`Spec`.
+- The validation logic in the API server is responsible for restricting the fields
+which can be edited in a StatefulSet specification.
+  - Part of the validation that the API server performs for objects (eg `StatefulSet`, `PVC`) happens in the object's API server code and part happens in an admission plugin.
```

Comment:
I don't recall any validation happening in admission, admission plugins are optional, all the validation we do is either inside the apiserver. Am I missing something?

Reply:
Is it the PVC validation that happens in admission plugins? I'll admit, I'm not familiar with that part of the code base :sweat_smile:

Reply:
Part of it happens here https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/persistentvolume/resize/admission.go 
Part of the validation that happens in that admission is making sure that the storage class supports expansions

Reply:
If you're saying this is "layout the various components involved in creating and managing the persistent volumes associated with `StatefulSet`" can you make sure this is properly described such that all readers of this proposal can follow what are the actors and what is the relationship between them.